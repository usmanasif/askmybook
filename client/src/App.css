.App {
  margin: 2rem;
  text-align: center;
  padding: calc(1vmin + 0.5rem);
  margin-inline: auto;
  font-size: clamp(1em, 0.909em + 0.45vmin, 1.25em);
  font-family: system-ui;
  max-width: 600px;
}

.book {
  height: 200px;
  box-shadow: 0 1px 2px rgb(0 0 0 / 20%);
  border-radius: 10px;
}

.title {
  font-weight: bold;
  margin-bottom: 0;
  font-size: 24px;
  line-height: 36px;
}

.credits {
  display: block;
  color: #777;
  text-align: left;
}

.question {
  box-sizing: border-box;
  width: 100%;
  border: 1px solid black;
  border-radius: 10px;
  font-size: 18px;
  padding: 8px 12px;
}

.lucky {
  background: #eee; 
  border-color: #eee; 
  color: #444;
}

.buttons {
  display: flex;
  justify-content: center;
  gap: 1rem;
}

button {
  display: inline-block;
  color: white;
  background: black;
  margin-top: 1rem;
  width: auto;
  padding: 8px 18px;
  line-height: 1.75;
  border: 1px solid black;
  border-radius: 10px;
  font-size: 18px;
}

.other-question {
  display: flex;
}

.hide {
  display: none;
}

.answer {
  text-align: left;
  line-height: 1.75;
}

.loading {
  cursor: not-allowed;
  opacity: 0.5;
}