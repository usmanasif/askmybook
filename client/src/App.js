import { useEffect, useState } from 'react'
import axios from 'axios'

import book from './book.png'
import './App.css'


const App = () => {
  // const ANSWER = 'The Minimalist Entrepreneur is a book about how to start and grow a business with less stress and fewer resources. It covers topics like how to choose what business to start, how to build and sell your product, and how to manage your time and money.'
  const question_id = null

  const [question, setQuestion] = useState('hello')
  const [answer, setAnswer] = useState('')
  const [letterIndex, setLetterIndex] = useState(0)
  const [loading, setLoading] = useState(false)
  let flag = false

  useEffect(() => {
    let intervalId = null
    if (letterIndex < answer.length) {
      intervalId = setInterval(() => {
        setLetterIndex(letterIndex + 1)
      }, 40)
    } else {
      clearInterval(intervalId)
    }
    return () => clearInterval(intervalId)
  }, [answer, letterIndex])

  useEffect(() => {
    if (flag) {
      return
    } else {
      flag = true
      const questionId = window.location.pathname.split('/')[2]
      
      if(typeof questionId !== 'undefined')
      fetchAnswer(questionId, null)
      
      return () => {
        if ('speechSynthesis' in window) {
          speechSynthesis.cancel()
        }
      }
    }
  }, [])

  const handleChange = ({target}) => {
    setQuestion(target.value)
  }

  const answerHanlder = () => {
    setLoading(true)
    fetchAnswer(null, question)
  }

  const fetchAnswer = (id, question) => {
    axios.post('http://localhost:3001/api/v1/questions', {
      id: id,
      title: question
    })
      .then(({data}) => {
        window.history.pushState({}, '', `/question/${data.id}`)
        setQuestion(data.question)
        data.question.length ? setAnswer(data.answer) : alert('Please ask a question!')
        setLoading(false)
        speechHanlder(data.answer)
      })
      .catch(err => {
        alert(err)
      })
  }

  const speechHanlder = answer => {
    if ('speechSynthesis' in window) {
      const utterance = new SpeechSynthesisUtterance(answer)
      utterance.rate = 0.9
      window.speechSynthesis.speak(utterance)
    }
  }

  const askQuestion = () => {
    setAnswer('')
    speechSynthesis.cancel()
    setLetterIndex(0)
  }

  return <div className='App'>
    <img src={book} className='book' alt='logo' />
    <h1 className='title'>Ask My Book</h1>
    <p className='credits'>
      This is an experiment in using AI to make my book's content more accessible. Ask a question and AI will answer it in real-time:
    </p>
    <textarea
      value={question}
      placeholder='Ask a question'
      onChange={handleChange}
      rows='4'
      className='question'
    />
    {answer.length === 0 ? 
      <div className='buttons'>
        <button 
          className={'' + (loading && 'loading')}
          onClick={answerHanlder}
          disabled={loading}>
            {loading ? 'Asking...' : 'Ask question'}
          </button>
        <button className='lucky' onClick={answerHanlder}>I'm feeling lucky</button>
      </div>
      : 
      <p className='answer'>
        <strong>Answer:</strong>
        <span> {answer.slice(0, letterIndex)} </span>
        {letterIndex === answer.length && 
          <button 
            className='other-question'
            onClick={askQuestion}
          >
            Ask another question
          </button>
        }
      </p>
    }
  </div>
}

export default App
