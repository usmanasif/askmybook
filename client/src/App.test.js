import React from 'react'
import { render, fireEvent } from '@testing-library/react'
import App from './App'
// import axios from '../__mocks__/axios';

// axios.__setMockResponse({
//   status: 200,
//   data: {
//     answer: 'The Minimalist Entrepreneur is a book about how to start and grow a business with less stress and fewer resources.',
//   },
// });

describe('App component', () => {
  test('renders App component', () => {
    const { container } = render(<App />)
    expect(container).toMatchSnapshot()
  })

  test('renders initial component', () => {
    const { asFragment } = render(<App />)
    expect(asFragment()).toMatchSnapshot()
  })

  it('renders the title and buttons', () => {
    const { getByText } = render(<App />)

    expect(getByText('Ask My Book')).toBeInTheDocument()
    expect(getByText('Ask question')).toBeInTheDocument()
    expect(getByText("I'm feeling lucky")).toBeInTheDocument()
  })

  it('updates the question when textarea value changes', () => {
    const { getByPlaceholderText } = render(<App />)
    const question = getByPlaceholderText('Ask a question')

    fireEvent.change(question, { target: { value: 'What is your name?' } })
    expect(question.value).toBe('What is your name?')
  })

  // it('shows answer when ask question button is clicked', () => {
  //   const { getByText, getByPlaceholderText } = render(<App />)
  //   const question = getByPlaceholderText('Ask a question')
  //   const askButton = getByText('Ask question')

  //   fireEvent.change(question, { target: { value: 'What is your name?' } })
  //   fireEvent.click(askButton)
  //   expect(getByText('Answer:')).toBeInTheDocument()
  // })
})