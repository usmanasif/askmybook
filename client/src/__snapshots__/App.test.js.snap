// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component renders App component 1`] = `
<div>
  <div
    class="App"
  >
    <img
      alt="logo"
      class="book"
      src="book.png"
    />
    <h1
      class="title"
    >
      Ask My Book
    </h1>
    <p
      class="credits"
    >
      This is an experiment in using AI to make my book's content more accessible. Ask a question and AI will answer it in real-time:
    </p>
    <textarea
      class="question"
      placeholder="Ask a question"
      rows="4"
    >
      hello
    </textarea>
    <div
      class="buttons"
    >
      <button
        class="false"
      >
        Ask question
      </button>
      <button
        class="lucky"
      >
        I'm feeling lucky
      </button>
    </div>
  </div>
</div>
`;

exports[`App component renders initial component 1`] = `
<DocumentFragment>
  <div
    class="App"
  >
    <img
      alt="logo"
      class="book"
      src="book.png"
    />
    <h1
      class="title"
    >
      Ask My Book
    </h1>
    <p
      class="credits"
    >
      This is an experiment in using AI to make my book's content more accessible. Ask a question and AI will answer it in real-time:
    </p>
    <textarea
      class="question"
      placeholder="Ask a question"
      rows="4"
    >
      hello
    </textarea>
    <div
      class="buttons"
    >
      <button
        class="false"
      >
        Ask question
      </button>
      <button
        class="lucky"
      >
        I'm feeling lucky
      </button>
    </div>
  </div>
</DocumentFragment>
`;
